// gitclaw/dist/audit.js
import { appendFile, mkdir } from "fs/promises";
import { join, dirname } from "path";
export class AuditLogger {
    logPath;
    sessionId;
    enabled;
    constructor(gitagentDir, sessionId, enabled) {
        this.logPath = join(gitagentDir, "audit.jsonl");
        this.sessionId = sessionId;
        this.enabled = enabled;
    }
    async log(event, data = {}) {
        if (!this.enabled)
            return;
        const entry = {
            timestamp: new Date().toISOString(),
            session_id: this.sessionId,
            event,
            ...data,
        };
        try {
            await mkdir(dirname(this.logPath), { recursive: true });
            await appendFile(this.logPath, JSON.stringify(entry) + "\n", "utf-8");
        }
        catch {
            // Audit logging failures are non-fatal
        }
    }
    async logToolUse(tool, args) {
        await this.log("tool_use", { tool, args });
    }
    async logToolResult(tool, result) {
        await this.log("tool_result", { tool, result: result.slice(0, 1000) });
    }
    async logResponse() {
        await this.log("response");
    }
    async logError(error) {
        await this.log("error", { error });
    }
    async logSessionStart() {
        await this.log("session_start");
    }
    async logSessionEnd() {
        await this.log("session_end");
    }
}
/**
 * Check if audit logging is enabled via compliance config.
 */
export function isAuditEnabled(compliance) {
    if (!compliance)
        return false;
    return compliance.recordkeeping?.audit_logging === true;
}
