// gitclaw/dist/plugin-sdk.js
export function createPluginApi(pluginId, pluginDir, config) {
    const tools = [];
    const hooks = {};
    const memoryLayers = [];
    let promptText = "";
    const prefix = `[plugin:${pluginId}]`;
    return {
        pluginId,
        pluginDir,
        config,
        registerTool(def) {
            tools.push(def);
        },
        registerHook(event, handler) {
            if (!hooks[event])
                hooks[event] = [];
            hooks[event].push(handler);
        },
        addPrompt(text) {
            promptText = promptText ? `${promptText}\n\n${text}` : text;
        },
        registerMemoryLayer(layer) {
            memoryLayers.push(layer);
        },
        logger: {
            info(msg) { console.log(`${prefix} ${msg}`); },
            warn(msg) { console.warn(`${prefix} ${msg}`); },
            error(msg) { console.error(`${prefix} ${msg}`); },
        },
        getTools() {
            return tools;
        },
        getHooks() {
            const events = Object.keys(hooks);
            if (events.length === 0)
                return null;
            const result = {};
            for (const event of events) {
                const handlers = hooks[event];
                // Wrap each programmatic handler as a HookDefinition
                // that uses a synthetic script path with inline execution
                result[event] = handlers.map((handler, i) => ({
                    script: `__programmatic_${pluginId}_${event}_${i}`,
                    description: `Programmatic hook from plugin ${pluginId}`,
                    baseDir: pluginDir,
                    _handler: handler, // attached for programmatic execution
                }));
            }
            return result;
        },
        getPrompt() {
            return promptText;
        },
        getMemoryLayers() {
            return memoryLayers;
        },
    };
}
