// gitclaw/dist/sandbox.js
import { execSync } from "child_process";
// ── Factory ─────────────────────────────────────────────────────────────
function detectRepoUrl(dir) {
    try {
        return execSync("git remote get-url origin", { cwd: dir, stdio: "pipe" })
            .toString()
            .trim();
    }
    catch {
        return null;
    }
}
/**
 * Create a SandboxContext by dynamically importing gitmachine.
 * Throws a clear error if gitmachine is not installed.
 */
export async function createSandboxContext(config, dir) {
    let gitmachine;
    try {
        // @ts-ignore — gitmachine is an optional peer dependency
        gitmachine = await import("gitmachine");
    }
    catch {
        throw new Error("Sandbox mode requires the 'gitmachine' package.\n" +
            "Install it with: npm install gitmachine");
    }
    const token = config.token
        || process.env.GITHUB_TOKEN
        || process.env.GIT_TOKEN;
    const repository = config.repository || detectRepoUrl(dir);
    if (!repository) {
        throw new Error("Sandbox mode requires a repository URL. Provide it via --sandbox config, " +
            "or ensure the working directory has a git remote named 'origin'.");
    }
    const gitMachine = new gitmachine.GitMachine({
        provider: config.provider,
        template: config.template,
        timeout: config.timeout,
        repository,
        token,
        session: config.session,
        autoCommit: config.autoCommit ?? true,
        envs: config.envs,
    });
    // The repo path inside the sandbox is determined by gitmachine after start().
    // Convention: /home/user/<repo-name>
    const repoName = repository.split("/").pop()?.replace(/\.git$/, "") || "repo";
    const repoPath = `/home/user/${repoName}`;
    return {
        gitMachine,
        machine: gitMachine.machine,
        repoPath,
    };
}
