// gitclaw/dist/tools/sandbox-cli.js
import { cliSchema, MAX_OUTPUT, DEFAULT_TIMEOUT, truncateOutput } from "./shared.js";
export function createSandboxCliTool(ctx, defaultTimeout) {
    const baseTimeout = defaultTimeout ?? DEFAULT_TIMEOUT;
    return {
        name: "cli",
        label: "cli",
        description: "Execute a shell command in the sandbox VM. Returns stdout and stderr combined. Output is truncated if it exceeds ~100KB. Default timeout is 120 seconds.",
        parameters: cliSchema,
        execute: async (_toolCallId, { command, timeout }, signal, onUpdate) => {
            if (signal?.aborted)
                throw new Error("Operation aborted");
            const timeoutSecs = timeout ?? baseTimeout;
            let output = "";
            const result = await ctx.gitMachine.run(command, {
                cwd: ctx.repoPath,
                timeout: timeoutSecs,
                onStdout: (data) => {
                    output += data;
                    if (onUpdate && output.length <= MAX_OUTPUT) {
                        onUpdate({
                            content: [{ type: "text", text: output }],
                            details: undefined,
                        });
                    }
                },
                onStderr: (data) => {
                    output += data;
                    if (onUpdate && output.length <= MAX_OUTPUT) {
                        onUpdate({
                            content: [{ type: "text", text: output }],
                            details: undefined,
                        });
                    }
                },
            });
            const exitCode = result?.exitCode ?? 0;
            let text = truncateOutput(output) || "(no output)";
            if (exitCode !== 0) {
                text += `\n\nExit code: ${exitCode}`;
                throw new Error(text);
            }
            return {
                content: [{ type: "text", text }],
                details: { exitCode },
            };
        },
    };
}
