// gitclaw/dist/learning/reinforcement.js
import { readFile, writeFile } from "fs/promises";
import { join } from "path";
import yaml from "js-yaml";
const MAX_NEGATIVE_EXAMPLES = 10;
const DEFAULT_STATS = {
    confidence: 1.0,
    usage_count: 0,
    success_count: 0,
    failure_count: 0,
    negative_examples: [],
};
// ── Confidence math ─────────────────────────────────────────────────────
export function adjustConfidence(current, outcome, failureReason) {
    const stats = { ...current };
    stats.usage_count++;
    switch (outcome) {
        case "success":
            // Asymptotic to 1.0: conf + 0.1 * (1 - conf)
            stats.confidence = Math.min(1.0, stats.confidence + 0.1 * (1 - stats.confidence));
            stats.success_count++;
            break;
        case "failure":
            // 2x penalty (asymmetric loss)
            stats.confidence = Math.max(0.0, stats.confidence - 0.2);
            stats.failure_count++;
            if (failureReason) {
                stats.negative_examples = [
                    ...stats.negative_examples.slice(-(MAX_NEGATIVE_EXAMPLES - 1)),
                    failureReason,
                ];
            }
            break;
        case "partial":
            stats.confidence = Math.max(0.0, stats.confidence - 0.05);
            stats.failure_count++;
            if (failureReason) {
                stats.negative_examples = [
                    ...stats.negative_examples.slice(-(MAX_NEGATIVE_EXAMPLES - 1)),
                    failureReason,
                ];
            }
            break;
    }
    // Round to avoid floating-point drift
    stats.confidence = Math.round(stats.confidence * 100) / 100;
    return stats;
}
// ── SKILL.md frontmatter read/write ─────────────────────────────────────
function parseFrontmatter(content) {
    const match = content.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$/);
    if (!match) {
        return { frontmatter: {}, body: content };
    }
    const frontmatter = yaml.load(match[1]);
    return { frontmatter, body: match[2] };
}
function serializeFrontmatter(frontmatter, body) {
    const yamlStr = yaml.dump(frontmatter, { lineWidth: -1, noRefs: true }).trimEnd();
    return `---\n${yamlStr}\n---\n${body}`;
}
export async function loadSkillStats(skillDir) {
    const skillFile = join(skillDir, "SKILL.md");
    try {
        const content = await readFile(skillFile, "utf-8");
        const { frontmatter } = parseFrontmatter(content);
        return {
            confidence: typeof frontmatter.confidence === "number" ? frontmatter.confidence : DEFAULT_STATS.confidence,
            usage_count: typeof frontmatter.usage_count === "number" ? frontmatter.usage_count : DEFAULT_STATS.usage_count,
            success_count: typeof frontmatter.success_count === "number" ? frontmatter.success_count : DEFAULT_STATS.success_count,
            failure_count: typeof frontmatter.failure_count === "number" ? frontmatter.failure_count : DEFAULT_STATS.failure_count,
            negative_examples: Array.isArray(frontmatter.negative_examples) ? frontmatter.negative_examples : [],
        };
    }
    catch {
        return { ...DEFAULT_STATS };
    }
}
export async function saveSkillStats(skillDir, stats) {
    const skillFile = join(skillDir, "SKILL.md");
    const content = await readFile(skillFile, "utf-8");
    const { frontmatter, body } = parseFrontmatter(content);
    frontmatter.confidence = stats.confidence;
    frontmatter.usage_count = stats.usage_count;
    frontmatter.success_count = stats.success_count;
    frontmatter.failure_count = stats.failure_count;
    frontmatter.negative_examples = stats.negative_examples;
    await writeFile(skillFile, serializeFrontmatter(frontmatter, body), "utf-8");
}
export function isSkillFlagged(stats) {
    return stats.confidence < 0.4;
}
