// gitclaw/dist/composio/client.js
// Composio REST API v3 client — zero dependencies, uses native fetch()
const BASE_URL = "https://backend.composio.dev/api/v3";
// ── Client ───────────────────────────────────────────────────────────
export class ComposioClient {
    apiKey;
    // Cache auth config IDs so we don't recreate them every connect
    authConfigCache = new Map();
    constructor(apiKey) {
        this.apiKey = apiKey;
    }
    // List available toolkits, optionally merging connection status for a user
    async listToolkits(userId) {
        const resp = await this.request("GET", "/toolkits");
        const toolkits = Array.isArray(resp) ? resp : (resp.items ?? resp.toolkits ?? []);
        let connectedSlugs = new Set();
        if (userId) {
            try {
                const conns = await this.listConnections(userId);
                connectedSlugs = new Set(conns.map((c) => c.toolkitSlug));
            }
            catch {
                // If connections fail, just show all as disconnected
            }
        }
        return toolkits.map((tk) => ({
            slug: tk.slug ?? "",
            name: tk.name ?? tk.slug ?? "",
            description: tk.meta?.description ?? tk.description ?? "",
            logo: tk.meta?.logo ?? tk.logo ?? "",
            authSchemes: tk.auth_schemes ?? [],
            noAuth: tk.no_auth ?? false,
            connected: connectedSlugs.has(tk.slug ?? ""),
        }));
    }
    // Search tools across connected toolkits by natural language query
    // Makes parallel per-toolkit requests since the API doesn't support comma-separated toolkit_slug with query
    async searchTools(query, toolkitSlugs, limit = 10) {
        const mapTool = (t) => ({
            name: t.name ?? t.enum ?? "",
            slug: t.slug ?? t.enum ?? t.name ?? "",
            description: t.description ?? "",
            toolkitSlug: t.toolkit?.slug ?? t.toolkit_slug ?? "",
            parameters: t.input_parameters ?? t.parameters ?? t.inputParameters ?? {},
        });
        if (!toolkitSlugs?.length) {
            const params = new URLSearchParams({ query, limit: String(limit) });
            const resp = await this.request("GET", `/tools?${params}`);
            const tools = Array.isArray(resp) ? resp : (resp.items ?? resp.tools ?? []);
            return tools.map(mapTool);
        }
        // Parallel per-toolkit search
        const perToolkit = await Promise.all(toolkitSlugs.map(async (slug) => {
            try {
                const params = new URLSearchParams({ query, toolkit_slug: slug, limit: String(limit) });
                const resp = await this.request("GET", `/tools?${params}`);
                const tools = Array.isArray(resp) ? resp : (resp.items ?? resp.tools ?? []);
                return tools.map(mapTool);
            }
            catch {
                return [];
            }
        }));
        return perToolkit.flat().slice(0, limit);
    }
    // List tools for a specific toolkit
    async listTools(toolkitSlug) {
        const resp = await this.request("GET", `/tools?toolkit_slug=${encodeURIComponent(toolkitSlug)}`);
        const tools = Array.isArray(resp) ? resp : (resp.items ?? resp.tools ?? []);
        return tools.map((t) => ({
            name: t.name ?? t.enum ?? "",
            slug: t.slug ?? t.enum ?? t.name ?? "",
            description: t.description ?? "",
            toolkitSlug,
            parameters: t.input_parameters ?? t.parameters ?? t.inputParameters ?? {},
        }));
    }
    // Get or create an auth config for a toolkit (needed before creating a connection)
    async getOrCreateAuthConfig(toolkitSlug) {
        // Check cache first
        const cached = this.authConfigCache.get(toolkitSlug);
        if (cached)
            return cached;
        // Check if one already exists
        const existing = await this.request("GET", `/auth_configs?toolkit_slug=${encodeURIComponent(toolkitSlug)}`);
        const items = existing.items ?? [];
        if (items.length > 0) {
            const id = items[0].id ?? items[0].auth_config?.id;
            if (id) {
                this.authConfigCache.set(toolkitSlug, id);
                return id;
            }
        }
        // Create a new one with Composio-managed auth
        const created = await this.request("POST", "/auth_configs", {
            toolkit: { slug: toolkitSlug },
            auth_scheme: "OAUTH2",
            use_composio_auth: true,
        });
        const id = created.auth_config?.id ?? created.id ?? "";
        if (id)
            this.authConfigCache.set(toolkitSlug, id);
        return id;
    }
    // Start OAuth connection flow (two-step: ensure auth config, then create connection)
    async initiateConnection(toolkitSlug, userId, redirectUrl) {
        const authConfigId = await this.getOrCreateAuthConfig(toolkitSlug);
        if (!authConfigId) {
            throw new Error(`Failed to get auth config for toolkit: ${toolkitSlug}`);
        }
        const body = {
            auth_config: { id: authConfigId },
            connection: {
                user_id: userId,
                ...(redirectUrl ? { callback_url: redirectUrl } : {}),
            },
        };
        const resp = await this.request("POST", "/connected_accounts", body);
        return {
            connectionId: resp.id ?? "",
            redirectUrl: resp.redirect_url ?? resp.redirect_uri ?? resp.redirectUrl ?? resp.redirectUri ?? "",
        };
    }
    // List active connections for a user
    async listConnections(userId) {
        const resp = await this.request("GET", `/connected_accounts?user_ids=${encodeURIComponent(userId)}&statuses=ACTIVE`);
        const items = Array.isArray(resp) ? resp : (resp.items ?? resp.connections ?? []);
        return items.map((c) => ({
            id: c.id ?? "",
            toolkitSlug: c.toolkit?.slug ?? c.toolkit_slug ?? c.appUniqueId ?? c.integrationId ?? "",
            status: c.status ?? "ACTIVE",
            createdAt: c.createdAt ?? c.created_at ?? "",
        }));
    }
    // Delete a connection
    async deleteConnection(id) {
        await this.request("DELETE", `/connected_accounts/${encodeURIComponent(id)}`);
    }
    // Execute a tool action
    async executeTool(toolSlug, userId, params, connectedAccountId) {
        const body = {
            arguments: params,
            user_id: userId,
        };
        if (connectedAccountId)
            body.connected_account_id = connectedAccountId;
        return this.request("POST", `/tools/execute/${encodeURIComponent(toolSlug)}`, body);
    }
    // ── Private ────────────────────────────────────────────────────────
    async request(method, path, body) {
        const url = `${BASE_URL}${path}`;
        const headers = {
            "x-api-key": this.apiKey,
            "Accept": "application/json",
        };
        if (body)
            headers["Content-Type"] = "application/json";
        const resp = await fetch(url, {
            method,
            headers,
            body: body ? JSON.stringify(body) : undefined,
        });
        if (!resp.ok) {
            const text = await resp.text().catch(() => "");
            throw new Error(`Composio API ${method} ${path} failed (${resp.status}): ${text}`);
        }
        if (resp.status === 204)
            return undefined;
        return resp.json();
    }
}
